### PR TITLE
saving data in DB and encrypting the password

### PR DESCRIPTION
**Pull request by Yosziad**
i also changed the login from get req to post req

### CODE REVIEW

File: server/controllers/order.js [hunk 1]
```diff
@@ -21,21 +21,37 @@ const newOrderValidation = (order) => {
...
+        foundOrders,
+      });
+    })
+    .catch((err) => {
+      logger.error(err);
+      return res.status(500).json({
+        message: err,
+      });
+    });
 };
```
Comment (beniostr) [inline server/controllers/order.js]: const getAllOrders = async (req, res) => {
  let foundOrders;
  try {
    foundOrders = await Order.find().exec();
  } catch (e) {
    logger.error(err);
    return res.status(500).json({
      message: err,
    });
  }
  
  if (!foundOrders) {
    return res.status(404).json({
      message: 'There are no orders to fetch',
    });
  }
  logger.info('Fetching all orders');
  return res.status(200).json({
    message: 'Orders details',
    foundOrders,
  });
};

File: server/controllers/order.js [hunk 2]
```diff
@@ -56,38 +71,56 @@ const createOrder = (req, res) => {
...
+      return res.status(200).json({
+        message: 'Order details',
+        foundOrder,
+      });
+    })
+    .catch((err) => {
+      logger.error(err);
+      return res.status(500).json({
+        message: err,
+      });
```
Comment (beniostr) [inline server/controllers/order.js]: use async await

File: server/controllers/order.js [hunk 3]
```diff
@@ -56,38 +71,56 @@ const createOrder = (req, res) => {
...
+      return res.status(200).json({
+        message: 'Orders details',
+        foundOrders,
+      });
+    })
+    .catch((err) => {
+      logger.error(err);
+      return res.status(500).json({
+        message: err,
+      });
```
Comment (beniostr) [inline server/controllers/order.js]: async await

File: server/controllers/product.js [hunk 4]
```diff
@@ -46,71 +46,103 @@ const createProduct = (req, res) => {
...
+        message: 'Products details',
+        foundProducts,
+      });
+    })
+    .catch((err) => {
+      logger.error(err);
+      return res.status(500).json({
+        message: err,
+      });
+    });
```
Comment (beniostr) [inline server/controllers/product.js]: async await

File: server/controllers/product.js [hunk 5]
```diff
@@ -46,71 +46,103 @@ const createProduct = (req, res) => {
...
+      return res.status(200).json({
+        message: 'Product details',
+        foundProduct,
+      });
+    })
+    .catch((err) => {
+      logger.error(err);
+      return res.status(500).json({
+        message: err,
+      });
```
Comment (beniostr) [inline server/controllers/product.js]: async await

File: server/controllers/product.js [hunk 6]
```diff
@@ -46,71 +46,103 @@ const createProduct = (req, res) => {
...
+      logger.info(`Product with ${id} id was updated`, result);
+      return res.status(200).json({
+        message: 'successfully updated the product',
+      });
+    })
+    .catch((err) => {
+      logger.error(err);
+      return res.status(500).json({
+        message: err,
+      });
```
Comment (beniostr) [inline server/controllers/product.js]: async await

File: server/controllers/product.js [hunk 7]
```diff
@@ -46,71 +46,103 @@ const createProduct = (req, res) => {
...
+      logger.info(`Product with ${id} id was deleted`, result);
+      return res.status(200).json({
+        message: 'successfully deleted the product',
+      });
+    })
+    .catch((err) => {
+      logger.error(err);
+      return res.status(500).json({
+        message: err,
+      });
```
Comment (beniostr) [inline server/controllers/product.js]: async await

File: server/controllers/user.js [hunk 8]
```diff
@@ -1,25 +1,28 @@
...
   const isUserNameValid = userNameValidationRegEx.test(user.userName);
   if (!isUserNameValid) {
     throw new Error('username is invalid');
   }
-  if (!user.password) {
+  if (!password) {
     throw new Error('password field is empty');
   }
-  if ((user.password).length < 8) {
+  if ((password).length < 8) {
```
Comment (beniostr) [inline server/controllers/user.js]: password.length < 8

File: server/controllers/user.js [hunk 9]
```diff
@@ -36,62 +39,88 @@ const newUserValidation = (user) => {
...
+        return res.status(500).json({
+          message: error,
+        });
+      });
+    })
+    .catch((err) => {
+      logger.error(err);
+      return res.status(500).json({
+        message: err,
+      });
```
Comment (beniostr) [inline server/controllers/user.js]: async await

File: server/controllers/user.js [hunk 10]
```diff
@@ -36,62 +39,88 @@ const newUserValidation = (user) => {
...
+      newUserValidation(user, req.body.password);
+    } catch (err) {
+      logger.error(err, user);
+      return res.status(400).send('Error: creating a new user failed');
+    }
+    user.save();
+    logger.info('user created successfully', user);
+    return res.send('user created successfully');
+  });
 };
```
Comment (beniostr) [inline server/controllers/user.js]: const createUser = (req, res) => {
  const currUser = {
    userName: req.body.userName,
    password: req.body.password,
    firstName: req.body.firstName,
    lastName: req.body.lastName,
    address: req.body.address,
    age: req.body.age,
  };
  logger.info('handling create a user request', currUser);
  try {
    newUserValidation(currUser);
  } catch (err) {
    logger.error(err, user);
    return res.status(400).send('Error: creating a new user failed');
  }
  
  bcrypt.hash(req.body.password, saltRounds, async (error, hash) => {
    const user = new User({...currUser, password: hash});
    try {
      await user.save();
    } catch (err) {
      logger.error(err);
      return res.status(500).json({
        message: err.message,
      });
    }
    logger.info('user created successfully', user);
    return res.send('user created successfully');
  });
};

File: server/controllers/user.js [hunk 11]
```diff
@@ -36,62 +39,88 @@ const newUserValidation = (user) => {
...
+      logger.info(`A user with ${id} id was deleted`, result);
+      return res.status(200).json({
+        message: 'successfully deleted the user',
+      });
+    })
+    .catch((err) => {
+      logger.error(err);
+      return res.status(500).json({
+        message: err,
+      });
```
Comment (beniostr) [inline server/controllers/user.js]: async await

**Review by beniostr — CHANGES_REQUESTED** [review]

**Comment by beniostr** [comment]
use tokens

router.post('/', controller.createUser); // guest + give token

router.post('/login', controller.loginUser); // guest + give token

router.delete('/:userId', controller.deleteUser); // user

router.post('/', controller.createProduct); // admin token

router.get('/', controller.getProducts); // guest

router.get('/:productId', controller.getProductById); // guest

router.patch('/:productId', controller.updateProduct);// admin

router.delete('/:productId', controller.deleteProduct);// admin

router.get('/', controller.getAllOrders); // admin

router.post('/', controller.createOrder); // user

router.get('/orders/:orderId', controller.getOrder); // user

router.get('/usersOrders/:user', controller.getOrdersByUsername); // user

File: server/controllers/order.js [hunk 12]
```diff
@@ -20,22 +20,44 @@ const newOrderValidation = (order) => {
...
+    foundOrders = await Order.find().exec();
+  } catch (err) {
+    logger.error(err);
+    return res.status(500).json({
+      message: err,
+    });
+  }
+
+  if (!foundOrders) {
+    return res.status(404).json({
```
Comment (benios) [inline server/controllers/order.js:37]: missing log

File: server/controllers/order.js [hunk 13]
```diff
@@ -20,22 +20,44 @@ const newOrderValidation = (order) => {
...
   const { body } = req;
-  const order = {
+  const order = new Order({
     userName: body.userName,
     product: body.product,
     date: Date.now(),
-  };
+  });
+
+  order.save();
```
Comment (benios) [inline server/controllers/order.js]: you created an order before validating, and you didn't handle it, it is a promise, either use then and catch or use await with try and catch

File: server/controllers/order.js [hunk 14]
```diff
@@ -55,39 +76,71 @@ const createOrder = (req, res) => {
...
+  let foundOrder;
+  try {
+    foundOrder = await Order.findById(id).exec();
+  } catch (err) {
+    logger.error(err);
+    return res.status(500).json({
+      message: err,
+    });
+  }
+  if (!foundOrder) {
```
Comment (benios) [inline server/controllers/order.js]: missing log

File: server/controllers/order.js [hunk 15]
```diff
@@ -55,39 +76,71 @@ const createOrder = (req, res) => {
...
+  let foundOrders;
+  try {
+    foundOrders = await Order.find({ userName: username }).sort([['date', -1]]).exec();
+  } catch (err) {
+    logger.error(err);
+    return res.status(500).json({
+      message: err.message,
+    });
+  }
+  if (!foundOrders) {
```
Comment (benios) [inline server/controllers/order.js:126]: log missing

File: server/controllers/order.js [hunk 16]
```diff
@@ -55,39 +76,71 @@ const createOrder = (req, res) => {
...
-  const usersOrders = orderModel.getOrdersByUsername(username);
-  if (!usersOrders) {
+  const orderUser = req.userData.username;
+
+  try {
+    userPermission(username, orderUser);
+  } catch (err) {
+    logger.error(err);
+    return res.status(400).send(err.message);
+  }
```
Comment (benios) [inline server/controllers/order.js]: I think it is better to create a new middleware for this check

File: server/controllers/product.js [hunk 17]
```diff
@@ -46,70 +46,98 @@ const createProduct = (req, res) => {
     logger.error(err, product);
     return res.status(400).send('Error: creating a new product failed');
   }
-
-  productModel.createProduct(product);
+  product.save();
```
Comment (benios) [inline server/controllers/product.js]: then & catch ?
await with try and catch?

File: server/controllers/product.js [hunk 18]
```diff
@@ -46,70 +46,98 @@ const createProduct = (req, res) => {
...
+  try {
+    foundProducts = await Product.find().exec();
+  } catch (err) {
+    logger.error(err);
+    return res.status(500).json({
+      message: err.message,
+    });
+  }
+  if (!foundProducts) {
+    return res.status(404).json({
```
Comment (benios) [inline server/controllers/product.js:73]: missing log

File: server/controllers/product.js [hunk 19]
```diff
@@ -46,70 +46,98 @@ const createProduct = (req, res) => {
...
+  try {
+    foundProduct = await Product.findById(id).exec();
+  } catch (err) {
+    logger.error(err);
+    return res.status(500).json({
+      message: err.message,
+    });
+  }
+  if (!foundProduct) {
     return res.status(404).json({
```
Comment (benios) [inline server/controllers/product.js:97]: missing log

File: server/controllers/product.js [hunk 20]
```diff
@@ -46,70 +46,98 @@ const createProduct = (req, res) => {
...
+  let result;
+  try {
+    result = await Product.update({ _id: id }, props).exec();
+  } catch (err) {
+    logger.error(err);
+    return res.status(500).json({
+      message: err.message,
+    });
+  }
+  if (!result) {
```
Comment (benios) [inline server/controllers/product.js:120]: missing log

File: server/controllers/product.js [hunk 21]
```diff
@@ -46,70 +46,98 @@ const createProduct = (req, res) => {
...
+  let result;
+  try {
+    result = await Product.remove({ _id: id }).exec();
+  } catch (err) {
+    logger.error(err);
+    return res.status(500).json({
+      message: err.message,
+    });
+  }
+  if (!result) {
```
Comment (benios) [inline server/controllers/product.js:143]: missing log

File: server/controllers/user.js [hunk 22]
```diff
@@ -1,7 +1,14 @@
...
 const bodyParser = require('body-parser');
-const userModel = require('../model/user');
+const mongoose = require('mongoose');
+const bcrypt = require('bcrypt');
+const jwt = require('jsonwebtoken');
+const User = require('../model/user');
 const Logger = require('../services/logger_services');
+const role = require('../helpers/role');
+
+const saltRounds = 10;
```
Comment (benios) [inline server/controllers/user.js]: const SALT_ROUNDS = 10;
a standard for using "global" page consts

File: server/controllers/user.js [hunk 23]
```diff
@@ -36,62 +43,133 @@ const newUserValidation = (user) => {
...
+      message: err.message,
     });
   }
-  logger.setLogData(user);
-  logger.info('You passed a user ID');
-  return res.status(200).json({
-    message: 'You passed a user ID',
-    user,
+
+  if (!foundUser) {
```
Comment (benios) [inline server/controllers/user.js:58]: missing log

File: server/controllers/user.js [hunk 24]
```diff
@@ -36,62 +43,133 @@ const newUserValidation = (user) => {
...
+  if (result === true) {
+    const token = jwt.sign({ username: foundUser.userName, userId: foundUser._id, role: foundUser.role }, process.env.JWT_KEY, { expiresIn: '1h' });
+    logger.info('logged in successfully', foundUser);
+    return res.status(200).json({
+      message: 'User details',
+      foundUser,
+      token,
+    });
+  }
+  return res.status(401).json({
```
Comment (benios) [inline server/controllers/user.js:82]: missing log

File: server/controllers/user.js [hunk 25]
```diff
@@ -36,62 +43,133 @@ const newUserValidation = (user) => {
...
+    logger.error(error);
+    return res.status(400).send(error.message);
+  }
+  if (isUserExist !== null) {
+    logger.error('Username exists');
+    return res.status(409).send('Username exists, please try a different username');
+  }
   try {
-    newUserValidation(user);
+    newUserValidation(currUser);
```
Comment (benios) [inline server/controllers/user.js]: I think validation be before the existing user check

File: server/controllers/user.js [hunk 26]
```diff
@@ -36,62 +43,133 @@ const newUserValidation = (user) => {
...
-    logger.error(err, user);
-    return res.status(400).send('Error: creating a new user failed');
+    logger.error(err, currUser);
+    return res.status(400).send(err.message);
   }
 
-  userModel.createUser(user);
-  logger.info('user created successfully', user);
-  return res.send('user created successfully');
+  bcrypt.hash(req.body.password, saltRounds, async (error, hash) => {
```
Comment (benios) [inline server/controllers/user.js]: change line to use await with try and catch

File: server/controllers/user.js [hunk 27]
```diff
@@ -36,62 +43,133 @@ const newUserValidation = (user) => {
...
-  if (!didUserDeleted) {
+const deleteUser = async (req, res) => {
+  const id = req.params.userId;
+  const { userId } = req.userData;
+  try {
+    deletePermission(id, userId);
+  } catch (err) {
+    logger.error(err);
+    return res.status(400).send(err.message);
+  }
```
Comment (benios) [inline server/controllers/user.js]: you could have use the middleware I suggested and spare this code

File: server/controllers/user.js [hunk 28]
```diff
@@ -36,62 +43,133 @@ const newUserValidation = (user) => {
...
+  let result;
+  try {
+    result = await User.deleteOne({ _id: id }).exec();
+  } catch (err) {
+    logger.error(err);
+    return res.status(500).json({
+      message: err,
+    });
+  }
+  if (!result) {
```
Comment (benios) [inline server/controllers/user.js:154]: missing log

File: server/middleware/check-auth.js [hunk 29]
```diff
@@ -0,0 +1,29 @@
...
+}
+
+function authUser(req, res, next) {
+  try {
+    const token = req.headers.authorization.split(' ')[1];
+    const decoded = jwt.verify(token, process.env.JWT_KEY);
+    req.userData = decoded;
+    return next();
+  } catch (err) {
+    return res.status(500).json({
```
Comment (benios) [inline server/middleware/check-auth.js:26]: add log

File: server/middleware/check-auth.js [hunk 30]
```diff
@@ -0,0 +1,29 @@
+const jwt = require('jsonwebtoken');
+
+function authRole(role) {
+  return (req, res, next) => {
+    if (req.userData.role === role || req.userData.role === 'Admin') {
+      return next();
+    }
+    res.status(401);
+    return res.send('Not allowed');
```
Comment (benios) [inline server/middleware/check-auth.js]: combine to one line

Comment (benios) [reply server/middleware/check-auth.js]: and log

File: server/model/order.js [hunk 31]
```diff
@@ -1,25 +1,11 @@
-const utils = require('../utils/utils');
+const mongoose = require('mongoose');
 
-const ordersList = [];
+const orderSchema = new mongoose.Schema({
+  userName: String,
+  product: String,
+  date: Number,
+});
```
Comment (benios) [inline server/model/order.js]: check on required fields

**Review by benios — CHANGES_REQUESTED** [review]

**Pull request merged**